File: Android V1.6/app/src/main/java/com/arcinteractive/spaces/ui/screens/files/FilesViewModel.kt
package com.arcinteractive.spaces.ui.screens.files

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.arcinteractive.spaces.data.media.EncryptedMediaService
import com.arcinteractive.spaces.data.model.EncryptedMediaMetadata
import com.arcinteractive.spaces.data.model.MediaType
import com.arcinteractive.spaces.data.model.MessageType
import com.arcinteractive.spaces.data.model.Space
import com.arcinteractive.spaces.data.model.SpaceFileItem
import com.arcinteractive.spaces.data.model.SpaceFolder
import com.arcinteractive.spaces.data.model.SpaceMedia
import com.arcinteractive.spaces.data.spaces.SpaceService
import com.google.firebase.firestore.ListenerRegistration
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.text.DateFormat

data class DocumentActionPayload(
    val filePath: String,
    val mimeType: String
)

enum class FilesSortOption(val label: String) {
    Date("Date"),
    Name("Name"),
    Size("Size"),
    Type("Type")
}

data class FilesUiState(
    val folders: List<SpaceFolder> = emptyList(),
    val allFiles: List<SpaceFileItem> = emptyList(),
    val isLoading: Boolean = false,
    val currentUserId: String? = null,
    val canUploadFiles: Boolean = false,
    val canManageAllFiles: Boolean = false,
    val searchText: String = "",
    val sortOption: FilesSortOption = FilesSortOption.Date,
    val selectedMedia: SpaceMedia? = null,
    val previewDocument: DocumentActionPayload? = null,
    val shareDocument: DocumentActionPayload? = null,
    val renameTargetFile: SpaceFileItem? = null,
    val pendingDeleteFile: SpaceFileItem? = null,
    val lastErrorMessage: String? = null,
    val lastInfoMessage: String? = null
) {
    val files: List<SpaceFileItem>
        get() {
            val query = searchText.trim()
            val filtered = allFiles.filter { file ->
                query.isEmpty() || file.name.contains(query, ignoreCase = true)
            }

            return when (sortOption) {
                FilesSortOption.Date -> filtered.sortedByDescending { it.createdAt }
                FilesSortOption.Name -> filtered.sortedBy { it.name.lowercase() }
                FilesSortOption.Size -> filtered.sortedByDescending { it.sizeBytes }
                FilesSortOption.Type -> filtered.sortedWith(
                    compareBy<SpaceFileItem> { it.typeDescription.lowercase() }
                        .thenBy { it.name.lowercase() }
                )
            }
        }
}

class FilesViewModel(
    private val space: Space,
    private val spaceService: SpaceService = SpaceService(),
    private val encryptedMediaService: EncryptedMediaService = EncryptedMediaService()
) : ViewModel() {
    private val _uiState = MutableStateFlow(FilesUiState())
    val uiState: StateFlow<FilesUiState> = _uiState.asStateFlow()

    private var folderListener: ListenerRegistration? = null
    private var fileListener: ListenerRegistration? = null
    private var hasLoadedFolders = false
    private var hasLoadedFiles = false

    fun startListeningIfNeeded(context: Context) {
        if (folderListener != null || fileListener != null) return
        _uiState.update {
            it.copy(
                isLoading = true,
                currentUserId = spaceService.currentUserId(context),
                lastErrorMessage = null
            )
        }

        viewModelScope.launch {
            val canManage = spaceService.canManageModules(context, space)
            val canUpload = spaceService.canPerform(context, space, com.arcinteractive.spaces.data.model.SpacePermission.UploadFiles)
            _uiState.update {
                it.copy(
                    canManageAllFiles = canManage,
                    canUploadFiles = canUpload
                )
            }
        }

        folderListener = spaceService.listenToFolders(
            context = context,
            space = space,
            listenerKey = "files.${space.id}.folders"
        ) { result ->
            result.onSuccess { folders ->
                hasLoadedFolders = true
                _uiState.update { current ->
                    current.copy(
                        folders = folders,
                        isLoading = !(hasLoadedFolders && hasLoadedFiles)
                    )
                }
            }.onFailure { error ->
                hasLoadedFolders = true
                _uiState.update { current ->
                    current.copy(
                        isLoading = !(hasLoadedFolders && hasLoadedFiles),
                        lastErrorMessage = error.localizedMessage ?: "Unable to load folders."
                    )
                }
            }
        }

        fileListener = spaceService.listenToFiles(
            context = context,
            space = space,
            listenerKey = "files.${space.id}.items"
        ) { result ->
            result.onSuccess { files ->
                hasLoadedFiles = true
                _uiState.update { current ->
                    current.copy(
                        allFiles = files,
                        isLoading = !(hasLoadedFolders && hasLoadedFiles)
                    )
                }
            }.onFailure { error ->
                hasLoadedFiles = true
                _uiState.update { current ->
                    current.copy(
                        isLoading = !(hasLoadedFolders && hasLoadedFiles),
                        lastErrorMessage = error.localizedMessage ?: "Unable to load files."
                    )
                }
            }
        }
    }

    fun updateSearchText(text: String) {
        _uiState.update { it.copy(searchText = text) }
    }

    fun updateSortOption(option: FilesSortOption) {
        _uiState.update { it.copy(sortOption = option) }
    }

    fun canManage(file: SpaceFileItem): Boolean {
        val state = _uiState.value
        return file.uploadedBy == state.currentUserId || state.canManageAllFiles
    }

    fun uploadFile(context: Context, fileUri: android.net.Uri) {
        if (!_uiState.value.canUploadFiles) return
        viewModelScope.launch {
            runCatching {
                spaceService.uploadFile(context, space, fileUri)
            }.onFailure { error ->
                _uiState.update {
                    it.copy(lastErrorMessage = error.localizedMessage ?: "Unable to upload this file.")
                }
            }
        }
    }

    fun openFile(context: Context, file: SpaceFileItem) {
        viewModelScope.launch {
            runCatching {
                if (file.isImage || file.isVideo) {
                    val metadata = EncryptedMediaMetadata(
                        mediaId = file.id,
                        mediaType = if (file.isVideo) MediaType.Video else MediaType.Photo,
                        storagePath = file.storagePath,
                        thumbnailStoragePath = null,
                        encryptionVersion = file.encryptionVersion,
                        nonce = file.nonceBase64,
                        thumbnailNonce = null,
                        mimeType = file.mimeType,
                        fileSize = file.sizeBytes,
                        createdAt = file.createdAt,
                        uploadedBy = file.uploadedBy
                    )
                    _uiState.update {
                        it.copy(
                            selectedMedia = SpaceMedia(
                                id = file.id,
                                spaceId = space.id,
                                type = if (file.isVideo) MessageType.Video else MessageType.Image,
                                mediaCategory = if (file.isVideo) "video" else "photo",
                                mediaType = if (file.isVideo) MediaType.Video else MediaType.Photo,
                                placeholderIconName = if (file.isVideo) "video" else "photo",
                                caption = file.name,
                                senderName = file.uploadedByName,
                                timestamp = file.createdAt?.let(::formatTimestamp).orEmpty(),
                                metadata = metadata
                            )
                        )
                    }
                } else {
                    val bytes = spaceService.downloadFileBytes(context, space, file)
                    val baseName = file.name.substringBeforeLast('.', file.id).ifBlank { file.id }
                    val tempFile = encryptedMediaService.shareFile(
                        context = context,
                        bytes = bytes,
                        fileName = baseName,
                        extension = file.fileExtension
                    )
                    _uiState.update {
                        it.copy(
                            previewDocument = DocumentActionPayload(tempFile.absolutePath, file.mimeType)
                        )
                    }
                }
            }.onFailure { error ->
                _uiState.update {
                    it.copy(lastErrorMessage = error.localizedMessage ?: "Unable to open this file.")
                }
            }
        }
    }

    fun shareFile(context: Context, file: SpaceFileItem) {
        viewModelScope.launch {
            runCatching {
                val bytes = spaceService.downloadFileBytes(context, space, file)
                val baseName = file.name.substringBeforeLast('.', file.id).ifBlank { file.id }
                val tempFile = encryptedMediaService.shareFile(
                    context = context,
                    bytes = bytes,
                    fileName = baseName,
                    extension = file.fileExtension
                )
                _uiState.update {
                    it.copy(
                        shareDocument = DocumentActionPayload(tempFile.absolutePath, file.mimeType)
                    )
                }
            }.onFailure { error ->
                _uiState.update {
                    it.copy(lastErrorMessage = error.localizedMessage ?: "Unable to share this file.")
                }
            }
        }
    }

    fun downloadFile(context: Context, file: SpaceFileItem) {
        viewModelScope.launch {
            runCatching {
                val bytes = spaceService.downloadFileBytes(context, space, file)
                encryptedMediaService.saveFileToDownloads(context, bytes, file.name, file.mimeType)
            }.onSuccess { message ->
                _uiState.update { it.copy(lastInfoMessage = message) }
            }.onFailure { error ->
                _uiState.update {
                    it.copy(lastErrorMessage = error.localizedMessage ?: "Unable to download this file.")
                }
            }
        }
    }

    fun beginRename(file: SpaceFileItem) {
        _uiState.update { it.copy(renameTargetFile = file) }
    }

    fun cancelRename() {
        _uiState.update { it.copy(renameTargetFile = null) }
    }

    fun renameFile(context: Context, file: SpaceFileItem, newName: String) {
        viewModelScope.launch {
            runCatching {
                spaceService.renameFile(context, space, file, newName)
            }.onSuccess {
                _uiState.update { it.copy(renameTargetFile = null) }
            }.onFailure { error ->
                _uiState.update {
                    it.copy(lastErrorMessage = error.localizedMessage ?: "Unable to rename this file.")
                }
            }
        }
    }

    fun requestDelete(file: SpaceFileItem) {
        _uiState.update { it.copy(pendingDeleteFile = file) }
    }

    fun cancelDelete() {
        _uiState.update { it.copy(pendingDeleteFile = null) }
    }

    fun deletePendingFile(context: Context) {
        val file = _uiState.value.pendingDeleteFile ?: return
        viewModelScope.launch {
            runCatching {
                spaceService.softDeleteFile(context, space, file)
            }.onSuccess {
                _uiState.update { it.copy(pendingDeleteFile = null) }
            }.onFailure { error ->
                _uiState.update {
                    it.copy(lastErrorMessage = error.localizedMessage ?: "Unable to delete this file.")
                }
            }
        }
    }

    fun clearSelectedMedia() {
        _uiState.update { it.copy(selectedMedia = null) }
    }

    fun consumePreviewDocument() {
        _uiState.update { it.copy(previewDocument = null) }
    }

    fun consumeShareDocument() {
        _uiState.update { it.copy(shareDocument = null) }
    }

    fun clearLastErrorMessage() {
        _uiState.update { it.copy(lastErrorMessage = null) }
    }

    fun clearLastInfoMessage() {
        _uiState.update { it.copy(lastInfoMessage = null) }
    }

    override fun onCleared() {
        folderListener?.remove()
        fileListener?.remove()
        super.onCleared()
    }
}

private fun formatTimestamp(date: java.util.Date): String {
    return DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.SHORT).format(date)
}

File: iOS/Spaces V1.5/ViewModels/FilesViewModel.swift
import Combine
import FirebaseFirestore
import Foundation
import SwiftUI
import UniformTypeIdentifiers

@MainActor
final class FilesViewModel: ObservableObject {
    enum SortOption: String, CaseIterable, Identifiable {
        case dateNewest = "Date"
        case name = "Name"
        case size = "Size"
        case type = "Type"

        var id: String { rawValue }
    }

    struct ExportPayload {
        let document: ExportFileDocument
        let defaultFilename: String
        let contentType: UTType
    }

    @Published private(set) var folders: [SpaceFolder] = []
    @Published private(set) var allFiles: [SpaceFileItem] = []
    @Published private(set) var isLoading = false
    @Published private(set) var currentUserID: String?
    @Published private(set) var canManageAllFiles = false
    @Published var errorMessage: String?
    @Published var searchText = ""
    @Published var sortOption: SortOption = .dateNewest
    @Published var selectedMedia: SpaceMedia?
    @Published var previewDocumentURL: URL?
    @Published var shareURL: URL?
    @Published var exportPayload: ExportPayload?
    @Published var renameTargetFile: SpaceFileItem?
    @Published var pendingDeleteFile: SpaceFileItem?
    @Published private(set) var canUploadFiles = false

    let space: Space
    private let spaceService: SpaceService
    private var folderListener: ListenerRegistration?
    private var fileListener: ListenerRegistration?
    private var hasLoadedFolders = false
    private var hasLoadedFiles = false

    init(space: Space, spaceService: SpaceService? = nil) {
        self.space = space
        self.spaceService = spaceService ?? SpaceService()
    }

    deinit {
        folderListener?.remove()
        fileListener?.remove()
    }

    var files: [SpaceFileItem] {
        let trimmedQuery = searchText.trimmingCharacters(in: .whitespacesAndNewlines)
        let filtered = allFiles.filter { file in
            trimmedQuery.isEmpty || file.name.localizedCaseInsensitiveContains(trimmedQuery)
        }

        switch sortOption {
        case .dateNewest:
            return filtered.sorted { lhs, rhs in
                (lhs.createdAt ?? .distantPast) > (rhs.createdAt ?? .distantPast)
            }
        case .name:
            return filtered.sorted { lhs, rhs in
                lhs.name.localizedCaseInsensitiveCompare(rhs.name) == .orderedAscending
            }
        case .size:
            return filtered.sorted { lhs, rhs in
                lhs.sizeBytes > rhs.sizeBytes
            }
        case .type:
            return filtered.sorted { lhs, rhs in
                if lhs.typeDescription == rhs.typeDescription {
                    return lhs.name.localizedCaseInsensitiveCompare(rhs.name) == .orderedAscending
                }
                return lhs.typeDescription.localizedCaseInsensitiveCompare(rhs.typeDescription) == .orderedAscending
            }
        }
    }

    var isEmpty: Bool {
        !isLoading && files.isEmpty && folders.isEmpty
    }

    func startListeningIfNeeded() {
        guard folderListener == nil, fileListener == nil else { return }
        isLoading = true
        currentUserID = spaceService.currentUserID()

        Task {
            canManageAllFiles = await spaceService.canManageModules(in: space)
            canUploadFiles = await spaceService.canPerform(.uploadFiles, in: space)
        }

        folderListener = spaceService.listenToFolders(in: space) { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(let folders):
                self.folders = folders
            case .failure(let error):
                self.errorMessage = error.localizedDescription
            }
            self.hasLoadedFolders = true
            self.finishLoadingIfNeeded()
        }

        fileListener = spaceService.listenToFiles(in: space) { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(let files):
                self.allFiles = files
            case .failure(let error):
                self.errorMessage = error.localizedDescription
            }
            self.hasLoadedFiles = true
            self.finishLoadingIfNeeded()
        }
    }

    func canManage(_ file: SpaceFileItem) -> Bool {
        file.uploadedBy == currentUserID || canManageAllFiles
    }

    func uploadFile(from fileURL: URL) async {
        guard canUploadFiles else { return }
        do {
            _ = try await spaceService.uploadFile(in: space, from: fileURL)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func open(_ file: SpaceFileItem) async {
        do {
            if file.isImage || file.isVideo {
                selectedMedia = makeMedia(from: file)
                return
            }

            let data = try await spaceService.downloadFileData(file, in: space)
            previewDocumentURL = try temporaryFileURL(for: file, data: data)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func share(_ file: SpaceFileItem) async {
        do {
            let data = try await spaceService.downloadFileData(file, in: space)
            shareURL = try temporaryFileURL(for: file, data: data)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func prepareDownload(_ file: SpaceFileItem) async {
        do {
            let data = try await spaceService.downloadFileData(file, in: space)
            exportPayload = ExportPayload(
                document: ExportFileDocument(data: data),
                defaultFilename: file.name,
                contentType: contentType(for: file)
            )
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func rename(_ file: SpaceFileItem, to newName: String) async {
        do {
            try await spaceService.renameFile(file, in: space, to: newName)
            renameTargetFile = nil
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func deletePendingFile() async {
        guard let pendingDeleteFile else { return }
        do {
            try await spaceService.softDeleteFile(pendingDeleteFile, in: space)
            self.pendingDeleteFile = nil
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func clearShareURL() {
        shareURL = nil
    }

    func clearPreviewDocument() {
        previewDocumentURL = nil
    }

    func clearExportPayload() {
        exportPayload = nil
    }

    private func finishLoadingIfNeeded() {
        if hasLoadedFolders && hasLoadedFiles {
            isLoading = false
        }
    }

    private func makeMedia(from file: SpaceFileItem) -> SpaceMedia {
        let metadata = EncryptedMediaMetadata(
            mediaId: file.id,
            mediaType: file.isVideo ? .video : .photo,
            storagePath: file.storagePath,
            thumbnailStoragePath: nil,
            encryptionVersion: file.encryptionVersion,
            nonce: file.nonceBase64,
            thumbnailNonce: nil,
            mimeType: file.mimeType,
            fileSize: Int(file.sizeBytes),
            width: nil,
            height: nil,
            duration: nil,
            createdAt: file.createdAt,
            uploadedBy: file.uploadedBy
        )

        return SpaceMedia(
            id: file.id,
            spaceID: space.id,
            type: file.isVideo ? .video : .image,
            mediaCategory: file.isVideo ? "video" : "photo",
            mediaType: file.isVideo ? .video : .photo,
            placeholderImageName: file.isVideo ? "video.fill" : "photo.fill",
            caption: file.name,
            senderName: file.uploadedByName,
            timestamp: file.timestamp,
            metadata: metadata
        )
    }

    private func temporaryFileURL(for file: SpaceFileItem, data: Data) throws -> URL {
        let fileName = ((file.name as NSString).deletingPathExtension).isEmpty ? file.id : (file.name as NSString).deletingPathExtension
        return try EncryptedMediaService().shareURL(
            for: data,
            suggestedFileName: fileName,
            pathExtension: file.fileExtension
        )
    }

    private func contentType(for file: SpaceFileItem) -> UTType {
        if #available(iOS 14.0, *),
           let type = UTType(filenameExtension: file.fileExtension) {
            return type
        }
        return .data
    }
}

struct ExportFileDocument: FileDocument {
    static var readableContentTypes: [UTType] { [.data] }

    var data: Data

    init(data: Data) {
        self.data = data
    }

    init(configuration: ReadConfiguration) throws {
        data = configuration.file.regularFileContents ?? Data()
    }

    func fileWrapper(configuration: WriteConfiguration) throws -> FileWrapper {
        FileWrapper(regularFileWithContents: data)
    }
}
